File: solutions/cpp/phone-number/1/phone_number.cpp
#include "phone_number.h"

#include <string>
#include <regex>
#include <stdexcept>
#include <iostream>

using namespace std;

namespace phone_number {

        bool phone_number::validate_clean(string no){
            unsigned int len = no.length();
            if(len > 11 || len <10)
                return false;
            else if(len == 11 && no.at(0)!='1')
                return false;
            else if(no.at(len-10)-'0'<2)
                return false;
            else if(no.at(len-7) - '0'<2)
                return false;
            else 
                return true;
        }   
    

        phone_number::phone_number (string input){
            regex non_digits("\\D");
            string clean = regex_replace(input, non_digits,"");
            unsigned int len = clean.length();
            
            /* regex not_allowed("[^\\d-+.()\\s]"); 
            if(regex_search(input, not_allowed)){
                throw domain_error("Punctuation not allowed");                          }*/
        
            if (!validate_clean(clean))
                throw domain_error("Invalid number");                       
            this->the_number = clean.substr(len-10);
        }

        string phone_number::number() {
            return the_number;
        }
    
}  // namespace phone_number

File: solutions/julia/phone-number/1/phone-number.jl
function is_valid_clean(str)
    len = length(str)
    len <=11 && len >=10 || return false
    if len == 11 && str[1]!='1'
        false
    elseif parse(Int, str[end-9])<2 || parse(Int, str[end-6])<2
        false
    else
        true
    end
end

function clean(phone_number)
    clean = replace(phone_number, r"\D"=>"")
    is_valid_clean(clean) || throw(ArgumentError("invalid number"))
    clean[end-9:end]
end
